src/lib.rs

```diff
@@ -59,49 +94,64 @@ impl Command {
             .into_iter()
             .map(|s| s.as_ref().to_string())
             .collect::<Vec<String>>();
-        Ok(Command {
+        Ok(T::insert_command(Command {
             name: name.as_ref().to_string(),
             command: command.as_ref().to_string(),
             args: converted_args,
             cur_dir: None,
             env: HashMap::new(),
-        })
+        }))
     }
 
-    pub fn full_cmd<S, C>(name: S, command_string: C) -> Result<Command, CommandError>
+    fn full_cmd<S, C>(name: S, command_string: C) -> Result<Self, CommandError>
     where
         S: AsRef<str>,
         C: AsRef<str>,
     {
         let (command, args) = parse_command_string(command_string)?;
-        Ok(Command {
+        Ok(T::insert_command(Command {
             name: name.as_ref().to_string(),
             command,
             args,
             cur_dir: None,
             env: HashMap::new(),
-        })
+        }))
     }
 
-    pub fn cur_dir<D>(mut self, dir: D) -> Self
+    fn cur_dir<D>(&mut self, dir: D) -> &mut Self
     where
         D: AsRef<Path>,
     {
-        self.cur_dir = Some(dir.as_ref().to_path_buf());
+        self.get_command_mut().cur_dir = Some(dir.as_ref().to_path_buf());
         self
     }
 
-    pub fn env<K, V>(mut self, key: K, val: V) -> Self
+    fn env<K, V>(&mut self, key: K, val: V) -> &mut Self
     where
         K: AsRef<str>,
         V: AsRef<str>,
     {
-        self.env
+        self.get_command_mut()
+            .env
             .insert(key.as_ref().to_string(), val.as_ref().to_string());
         self
     }
 }
 
+impl CommandLike for Command {
+    fn insert_command(cmd: Command) -> Self {
+        cmd
+    }
```

Comment:
What does this function do? It doesn't seem to do anything from my point of view.

Reply:
It is used internally to construct Command and ConsoleCommand instances

Reply:
Hmm, but all it does is return its argument. And you only call it like this:

```rust
Ok(T::insert_command(Command {
            name: name.as_ref().to_string(),
            command: command.into(),
            args: converted_args,
            cur_dir: None,
            env: HashMap::new(),
        }))
```


What am I missing?

Reply:
Currently threr are 3 APIs, the channel API (base API), standard out API, and the writer API. 
They all take command-like objects (standard out api takes a command plus a few extra optional details like coloring and whatnot).

In order to unify the commands, there are CommandLike and CommandOperations traits to reduce code duplication and keep the APIs consistent.

This is in the impl block for CommandLike for Command. Of course, it is going to have a trivial implementation for CommandLike since it is exactly a command. Where things get slightly more interesting is in the impl block for ConsoleCommand where the command needs to get embedded in the ConsoleCommand wrapper struct.
```
    fn insert_command(cmd: Command) -> Self {
        ConsoleCommand {
            inner_command: cmd,
            color: Color::Random,
        }
    }
```
We can discuss this structure further on a call if we want? there might be a simpler way to structure it?

Reply:
Possible use of inner API in outer API that uses generics instead of encapsulation. This is really just an idea and might not work in this context. I really like your encapsulation approach as well.

```rs
// inner API
pub struct Command<O> {
    opts: O
}

// outer API
pub struct ConsoleOptions {
    color: Color,
}

pub type ConsoleCommand = Command<ConsoleOptions>;
```